agent/connectionuptime/service.go

```diff
@@ -86,7 +86,7 @@ func (c *Service) deleteOldEvents() {
 	expired := time.Now().Add(-c.windowPeriod)
 	for i := len(c.events) - 1; i >= 0; i-- {
 		if c.events[i].Timestamp.Before(expired) {
-			c.events[i].Timestamp = expired.Add(time.Second)
+			c.events[i].Timestamp = expired
```

Comment:
Is there a reason we need to rewrite the event's timestamp to `expired`?

Reply:
yes, there is

Imagine case when we have the next set of events
```[{time: 00:00, connected: true}, {time: 12:30, connected: false}, {time: 12:40, connected: true}]```

and you want to get a connection uptime during last hour from 12:40. In this case first event will be out of the period window (1 hour). 

In case if we only remove expired events then we lose information about connection between 00:00 and 12:30. And instead of removing we update (shift) the last expired timestamp to the `time.Now() - windowPeriod` time. 

And in our case events for calculation uptime will look like this: ```[{time: 11:40, connected: true}, {time: 12:30, connected: false}, {time: 12:40, connected: true}]```

Reply:
Thank you, that's clear now.
In these cases I usually recommend working with the window period instead of changing a timestamp. Changing the timestamp may have other unexpected consequences. The "truth" has changed.
When someone updates this part of code, they need to be aware the timestamp can change.

Imagine this situation:
1. `[{time: 00:00, connected: true}, {time: 12:30, connected: false}, {time: 12:40, connected: true}]`
2. Remove expired -> `[{time: 11:40, connected: true}, {time: 12:30, connected: false}, {time: 12:40, connected: true}]`
3. Set window period to 10 minutes
4. At 12:40 you would expect the uptime to be 0% whereas it would be >0% because we still have the 11:40 entry.

When calculating uptime, if we limited the total time to the window period instead, we could avoid this issue.
There could be more cases where this could potentially cause problems, if we relied on changing the timestamp.

I may be missing something, but I think working with the window period is less error-prone.

Reply:
> When calculating uptime, if we limited the total time to the window period instead, we could avoid this issue.

But window period can be changed only by user who starts pmm-agent. It's defined in the configuration.

Let's take a look on our example but without changing window period:
1. We have next events [{time: 00:00, connected: true}, {time: 12:30, connected: false}, {time: 12:40, connected: true}]
1. Remove expired event (which happened more than 1 hour before)-> [{time: 12:30, connected: false}, {time: 12:40, connected: true}

In this case first event was removed and we lost info what connections status was between 11:40 and 12:30.

Reply:
Based on the code the window period can be [changed here](https://github.com/percona/pmm/blob/0601387fb8f62e82275507b0e15bdce7d9e127bc/agent/commands/run.go#L77). 

Not sure if it can be really changed during runtime though.
In my example I don't recommend removing events, rather limiting the uptime calculation to the window period.
This means limiting the earliest time we account for in the [uptime calculation](https://github.com/percona/pmm/blob/0601387fb8f62e82275507b0e15bdce7d9e127bc/agent/connectionuptime/service.go#L156-L166).

If you had the following events
`[{time: 00:00, connected: true}, {time: 12:30, connected: false}, {time: 12:40, connected: true}]`
The current time is 12:40 and period is 1h.
The first 00:00 event would be considered only in the period 11:40-12:30.
This way having expired events does not cause issues with calculated uptime and there's no need to touch the timestamp.

Reply:
But without removing events at one point of time user will throw error that it's out of memory, we can forever add elements to slice without removing elements. Otherwise it will be memory leak here.

>If you had the following events
[{time: 00:00, connected: true}, {time: 12:30, connected: false}, {time: 12:40, connected: true}]
The current time is 12:40 and period is 1h.
The first 00:00 event would be considered only in the period 11:40-12:30.

But to calculate this we would need to know as well from which time we can calculate connection uptime (in our case it's 11:40). I don't see much difference between what you suggest and what's already implemented. Only in my implementation we update timestamp of last expired event from which we can start to calculate connection uptime. But I agree that it's hard to understand there.

In this case instead of updating existing event's timestamp, we can remove it and set a new event with updated timestamp. wdyt about this?

Reply:
>In this case instead of updating existing event's timestamp, we can remove it and set a new event with updated timestamp to make code more readable. wdyt about this?

@michal-kralik I pushed these changes, please take a look on it

Reply:
I agree we need to clear the expired events because of memory.
I suggest we change the way [uptime is calculated](https://github.com/percona/pmm/blob/1258affe0cb100984eeb9fc70ce3a39c2f3b7ff7/agent/connectionuptime/service.go#L159-L167).

Here's a pseudo-code I would suggest:

```go
expired := now - windowPeriod
for i, event := range c.events {
	if event.Connected {
		from := max(expired, event.timestamp)
		to := toTime
		if i+1 < len(c.events) {
			to = max(expired, c.events[i+1])
		}

		if from >= to {
			continue
		}

		connectedTime += to - from
	}
}
```

This way the calculation properly takes into account the window period only. And we can keep the timestamps intact.
Let me know if you think this makes sense.

Reply:
Hey Michal,

I got your point, it makes sense.
I changed implementation, please take a look on it